--- vte-0.11.13/src/vte.c~	2006-06-16 19:11:05.000000000 +0300
+++ vte-0.11.13/src/vte.c	2007-10-05 15:43:56.000000000 +0300
@@ -913,7 +913,10 @@
 			     &blink_cycle, NULL);
 	}
 
-	if (terminal->pvt->cursor_blink_timeout != blink_cycle) {
+	if (!terminal->pvt->cursor_blinks) {
+		terminal->pvt->cursor_blink_tag = 0;
+		return FALSE;
+	} else if (terminal->pvt->cursor_blink_timeout != blink_cycle) {
 		terminal->pvt->cursor_blink_tag = g_timeout_add_full(G_PRIORITY_LOW,
 								     blink_cycle / 2,
 								     vte_invalidate_cursor_periodic,
@@ -12170,11 +12173,13 @@
 		g_object_get(G_OBJECT(settings), "gtk-cursor-blink-time",
 			     &blink_cycle, NULL);
 	}
-	terminal->pvt->cursor_blink_tag = g_timeout_add_full(G_PRIORITY_LOW,
-							     blink_cycle / 2,
-							     vte_invalidate_cursor_periodic,
-							     terminal,
-							     NULL);
+	if (terminal->pvt->cursor_blinks) {
+		terminal->pvt->cursor_blink_tag = g_timeout_add_full(G_PRIORITY_LOW,
+								     blink_cycle / 2,
+								     vte_invalidate_cursor_periodic,
+								     terminal,
+								     NULL);
+	}
 
 	/* Set up input method support.  FIXME: do we need to handle the
 	 * "retrieve-surrounding" and "delete-surrounding" events? */
@@ -15228,6 +15233,24 @@
 {
 	g_return_if_fail(VTE_IS_TERMINAL(terminal));
 	terminal->pvt->cursor_blinks = blink;
+
+	if (blink && !terminal->pvt->cursor_blink_tag && GTK_WIDGET_REALIZED(terminal)) {
+		gint blink_cycle = 1000;
+		GtkSettings *settings = gtk_widget_get_settings(GTK_WIDGET(terminal));
+		if (G_IS_OBJECT(settings)) {
+			g_object_get(G_OBJECT(settings), "gtk-cursor-blink-time",
+					&blink_cycle, NULL);
+		}
+
+		terminal->pvt->cursor_blink_tag = g_timeout_add_full(G_PRIORITY_LOW,
+								     blink_cycle / 2,
+								     vte_invalidate_cursor_periodic,
+								     terminal,
+								     NULL);
+	} else if (!blink && terminal->pvt->cursor_blink_tag) {
+		g_source_remove(terminal->pvt->cursor_blink_tag);
+		terminal->pvt->cursor_blink_tag = 0;
+	}
 }
 
 /**
