diff -Nur vte-0.11.15/python/vte.defs vte-0.11.15.new/python/vte.defs
--- vte-0.11.15/python/vte.defs	2004-04-29 22:41:55.000000000 -0400
+++ vte-0.11.15.new/python/vte.defs	2005-10-28 15:59:50.000000000 -0400
@@ -7,6 +7,14 @@
   (gtype-id "VTE_TYPE_TERMINAL")
 )
 
+(define-object Reaper
+  (in-module "Vte")
+  (parent "GObject")
+  (c-name "VteReaper")
+  (gtype-id "VTE_TYPE_REAPER")
+)
+
+
 ;; Enumerations and flags ...
 
 (define-enum TerminalEraseBinding
@@ -665,3 +673,13 @@
 )
 
 
+;; From ../src/reaper.h
+(define-function vte_reaper_get_type
+  (c-name "vte_reaper_get_type")
+  (return-type "GtkType")
+)
+
+(define-function reaper_get
+  (c-name "vte_reaper_get")
+  (return-type "VteReaper")
+)
diff -Nur vte-0.11.15/python/vte.override vte-0.11.15.new/python/vte.override
--- vte-0.11.15/python/vte.override	2005-10-28 15:59:45.000000000 -0400
+++ vte-0.11.15.new/python/vte.override	2005-10-28 15:59:50.000000000 -0400
@@ -6,12 +6,14 @@
 #include <pygobject.h>
 #include <gtk/gtk.h>
 #include "../src/vte.h"
+#include "../src/reaper.h"
 %%
 modulename vte
 %%
 import gtk.gdk.Pixbuf as PyGdkPixbuf_Type
 import gtk.MenuShell as PyGtkMenuShell_Type
 import gtk.Widget as PyGtkWidget_Type
+import gobject.GObject as PyGObject_Type
 %%
 override vte_terminal_feed kwargs
 static PyObject *
diff -Nur vte-0.11.15/python/vtemodule.c vte-0.11.15.new/python/vtemodule.c
--- vte-0.11.15/python/vtemodule.c	2005-03-17 01:43:44.000000000 -0500
+++ vte-0.11.15.new/python/vtemodule.c	2005-10-28 15:59:50.000000000 -0400
@@ -25,6 +25,7 @@
 #include <pygobject.h>
 #include <pygtk/pygtk.h>
 #include "../src/vte.h"
+#include "../src/reaper.h"
 
 extern void pyvte_register_classes(PyObject * d);
 extern PyMethodDef pyvte_functions[];
